splink/comparison_level_library.py

```diff
@@ -68,3 +70,58 @@ def create_sql(self, sql_dialect: SplinkDialect) -> str:
 
     def create_label_for_charts(self) -> str:
         return f"Levenshtein distance of {self.col_name} <= {self.distance_threshold}"
+
+
+class DatediffLevel(ComparisonLevelCreator):
+    def __init__(
+        self,
+        col_name: str,
+        date_threshold: int,
+        date_metric: str = "day",  ##TODO: Lock down to sqlglot supported values
+        cast_strings_to_date: bool = False,
+        date_format: str = "%Y-%m-%d",
+    ):
+        """A comparison level using a date difference function
+
+        e.g. abs(date_diff('day', "mydate_l", "mydate_r")) <= 2  (duckdb dialect)
+
+        Args:
+            col_name (str): Input column name
+            date_threshold (int): The threshold for the date difference
+            date_metric (str): The unit of time ('day', 'month', 'year')
+                for the threshold
+            cast_strings_to_date (bool): Whether to cast string columns to date format
+            date_format (str): The format of the date string
+        """
+        super().__init__(col_name)
+        self.date_threshold = date_threshold
+        self.date_metric = date_metric
+        self.cast_strings_to_date = cast_strings_to_date
+        self.date_format = date_format
+
+    def create_sql(self, sql_dialect: SplinkDialect) -> str:
+        sqlglot_dialect_name = sql_dialect.sqlglot_name()
+        date_col = self.input_column(sql_dialect)
+        date_col_l, date_col_r = date_col.names_l_r()
+        col_l_no_dialect = parse_one(date_col_l, read=sqlglot_dialect_name).sql()
+        col_r_no_dialect = parse_one(date_col_r, read=sqlglot_dialect_name).sql()
+
+        if self.cast_strings_to_date:
+            col_l_no_dialect = f'STR_TO_TIME({col_l_no_dialect}, "{self.date_format}")'
+            col_r_no_dialect = f'STR_TO_TIME({col_r_no_dialect}, "{self.date_format}")'
+
+        undialected_sql = (
+            f"abs(date_diff({col_l_no_dialect}, "
```

Comment:
I think `date_diff` needs to be specified in our base classes - https://github.com/moj-analytical-services/splink/blob/splink4_dev/splink/dialects.py#L29.

Reply:
Ah, no, I understand why you're doing this.

The main reason I'd be tempted to stay away from sqlglot for this is due to:
* Error parsing.
* Not all functions are logged against each backend.

Ignoring that, this is a pattern we could more widely adopt across our backends. If we write the bulk of the class using standardised ANSI SQL and then grab differing SQL functions from sqlglot's databank.

Reply:
I think in my recent changes I've solved the issue of 'Not all functions are logged against each backend.'

Could you expand a bit on what you mean by 'error parsing'?

Reply:
Haha, sorry, I'm not sure why I called it 'error parsing'.

I simply meant that's it much harder to process the error effectively if we use sqlglot for this.

Either:
1. SQLglot errors stating that the sql can't be parsed or there's another confusing error.
2. It transpiles the SQL without any warnings and we get an error further down the line when we attempt to execute against a given backend.

Your solution should mean that this is no longer a problem.